File: src/main/java/model/GameData.java
package model;

import model.moves.Move;
import model.pieces.King;
import model.pieces.Piece;
import model.util.BoardMap;
import model.util.Colour;
import model.util.Position;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.*;

/**
 * Represents the board state (piece's positions) and the pieces that were eaten
 */
public class GameData implements Serializable {
    @NotNull
    private final BoardMap board;

    @NotNull
    private final EnumMap<Colour, Stack<Piece>> eatenPieces = new EnumMap<>(Colour.class);

    /**
     * A list of completed moves
     */
    private final LinkedList<Move> pastMoves = new LinkedList<>();

    /**
     * The kings for each player.
     */
    @NotNull
    private final EnumMap<Colour, King> kings = new EnumMap<>(Colour.class);

    public GameData(@NotNull BoardMap boardMap) {
        this.board = boardMap;

        for (Colour colour : Colour.values()) {
            eatenPieces.put(colour, new Stack<>());
        }

        for (Piece piece : boardMap.iteratePieces()) {
            if (piece instanceof King) {
                if (kings.containsKey(piece.getColour()))
                    throw new RuntimeException("There are two kings for the same player");

                kings.put(piece.getColour(), (King) piece);
            }
        }
    }

    @NotNull
    public BoardMap getBoard() {
        return board;
    }

    /**
     * @return the king of this colour
     */
    @NotNull
    King getKing(Colour colour) {
        return kings.get(colour);
    }

    /**
     * @return a collection of possible legal moves that can be player
     */
    @NotNull
    public Collection<Move> getPossibleLegalMoves(Colour colour) {
        return filterOnlyLegal(getAllPossibleMoves(colour), colour);
    }

    /**
     * @param moves        a collection of legal and illegal moves
     * @param verifierPour legal for who (will remove all moves that throw this colour's king in check)
     * @return a filter collection containing only legal moves
     */
    @NotNull
    public Collection<Move> filterOnlyLegal(@NotNull Collection<Move> moves, Colour verifierPour) {
        Collection<Move> legalMoves = new ArrayList<>();

        //For each move apply it and check if the king is in check
        for (Move move : moves) {
            move.apply(this);

            if (!isPieceAttacked(kings.get(verifierPour))) {
                legalMoves.add(move);
            }

            move.undo(this);
        }

        return legalMoves;
    }

    /**
     * @return true if the piece is being attacked by another piece
     */
    public boolean isPieceAttacked(@NotNull Piece piece) {
        return isPositionAttacked(board.getPosition(piece), piece.getColour() == Colour.WHITE ? Colour.BLACK : Colour.WHITE);
    }

    public boolean isPositionAttacked(Position position, Colour byWho) {
        for (Piece attacker : board.iteratePieces()) {
            //If piece opposite color check if attacks position
            if (attacker.getColour() == byWho && attacker.isAttackingPosition(this, position)) {
                return true;
            }
        }

        return false;
    }

    @NotNull
    private Set<Move> getAllPossibleMoves(Colour colour) {
        Set<Move> moves = new HashSet<>();

        for (Piece piece : board.iteratePieces()) {
            if (piece.getColour() == colour) {
                moves.addAll(piece.generatePossibleMoves(this, board.getPosition(piece)));
            }
        }
        return moves;
    }

    /**
     * @return the stack of eaten pieces for this colour
     */
    public Stack<Piece> getEatenPieces(Colour colour) {
        return eatenPieces.get(colour);
    }

    @NotNull
    public LinkedList<Move> getPastMoves() {
        return pastMoves;
    }
}

File: src/main/java/model/GameData.kt
package model

import model.moves.Move
import model.pieces.King
import model.pieces.Piece
import model.util.BoardMap
import model.util.Colour
import model.util.Position

import java.io.Serializable
import java.util.*

/**
 * Represents the board state (piece's positions) and the pieces that were eaten
 */
class GameData(val board: BoardMap) : Serializable {

    private val eatenPieces = EnumMap<Colour, Stack<Piece>>(Colour::class.java)

    /**
     * A list of completed moves
     */
    internal val pastMoves = LinkedList<Move>()

    /**
     * The kings for each player.
     */
    private val kings = EnumMap<Colour, King>(Colour::class.java)

    init {

        for (colour in Colour.values()) {
            eatenPieces[colour] = Stack()
        }

        for (piece in board.iteratePieces()) {
            if (piece is King) {
                if (kings.containsKey(piece.colour))
                    throw RuntimeException("There are two kings for the same player")

                kings[piece.colour] = piece
            }
        }
    }

    /**
     * @return the king of this colour
     */
    internal fun getKing(colour: Colour): King {
        return kings[colour]!!
    }

    /**
     * @param colour will not allow any moves that throw this colour's king in check
     * @return a collection of possible legal moves that can be player
     */
    internal fun getPossibleLegalMoves(colour: Colour): Collection<Move> {
        return filterOnlyLegal(getAllPossibleMoves(colour), colour)
    }

    /**
     * @param moves        a collection of legal and illegal moves
     * @param verifierPour legal for who (will remove all moves that throw this colour's king in check)
     * @return a filter collection containing only legal moves
     */
    internal fun filterOnlyLegal(moves: Collection<Move>, verifierPour: Colour): Collection<Move> {
        val legalMoves = ArrayList<Move>()

        //For each move apply it and check if the king is in check
        for (move in moves) {
            move.apply(this)

            if (!isPieceAttacked(kings[verifierPour]!!)) {
                legalMoves.add(move)
            }

            move.undo(this)
        }

        return legalMoves
    }

    /**
     * @return true if the piece is being attacked by another piece
     */
    internal fun isPieceAttacked(piece: Piece): Boolean {
        return isPositionAttacked(board.getPosition(piece), if (piece.colour == Colour.WHITE) Colour.BLACK else Colour.WHITE)
    }

    fun isPositionAttacked(position: Position?, byWho: Colour): Boolean {
        for (attacker in board.iteratePieces()) {
            //If piece opposite color check if attacks position
            if (attacker.colour == byWho && attacker.isAttackingPosition(this, position!!)) {
                return true
            }
        }

        return false
    }

    private fun getAllPossibleMoves(colour: Colour): Set<Move> {
        val moves = HashSet<Move>()

        for (piece in board.iteratePieces()) {
            if (piece.colour == colour) {
                moves.addAll(piece.generatePossibleMoves(this, board.getPosition(piece)!!))
            }
        }
        return moves
    }

    /**
     * @return the stack of eaten pieces for this colour
     */
    fun getEatenPieces(colour: Colour): Stack<Piece> {
        return eatenPieces[colour]!!
    }
}